File: frontend/src/app/core/services/email-account.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { EmailAccount, EmailAccountRequest } from '../../models/email-account.model';

/**
 * Handles CRUD operations for user email accounts, including IMAP/SMTP connection
 * testing and setting a default account.
 */
@Injectable({ providedIn: 'root' })
export class EmailAccountService {
  private readonly api = inject(ApiService);
  private readonly basePath = '/email-accounts';

  list() {
    return this.api.get<EmailAccount[]>(this.basePath);
  }

  get(id: string) {
    return this.api.get<EmailAccount>(`${this.basePath}/${id}`);
  }

  create(request: EmailAccountRequest) {
    return this.api.post<EmailAccount>(this.basePath, request);
  }

  update(id: string, request: EmailAccountRequest) {
    return this.api.put<EmailAccount>(`${this.basePath}/${id}`, request);
  }

  delete(id: string) {
    return this.api.delete<void>(`${this.basePath}/${id}`);
  }

  setDefault(id: string) {
    return this.api.patch<EmailAccount>(`${this.basePath}/${id}/default`, {});
  }

  testConnection(data: { host: string; port: number; username: string; password: string; ssl: boolean; type: 'imap' | 'smtp' }) {
    return this.api.post<{ success: boolean; message: string }>(`${this.basePath}/test-connection`, data);
  }
}

File: backend/src/main/java/com/postwerk/service/EmailAccountService.java
package com.postwerk.service;

import com.postwerk.dto.EmailAccountRequest;
import com.postwerk.dto.EmailAccountResponse;
import com.postwerk.dto.FolderResponse;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for managing user email accounts with IMAP/SMTP configuration,
 * including CRUD operations, default account selection, and IMAP folder management.
 *
 * @since 1.0
 */
public interface EmailAccountService {

    EmailAccountResponse create(UUID organizationId, UUID actingUserId, EmailAccountRequest request, String ipAddress);

    List<EmailAccountResponse> listByOrg(UUID organizationId);

    EmailAccountResponse getById(UUID organizationId, UUID accountId);

    EmailAccountResponse update(UUID organizationId, UUID actingUserId, UUID accountId, EmailAccountRequest request, String ipAddress);

    void delete(UUID organizationId, UUID actingUserId, UUID accountId, String ipAddress);

    EmailAccountResponse setDefault(UUID organizationId, UUID accountId);

    List<FolderResponse> listFolders(UUID organizationId, UUID accountId);

    FolderResponse createFolder(UUID organizationId, UUID accountId, String folderName);

    void deleteFolder(UUID organizationId, UUID accountId, UUID folderId);
}
